File: DSA/SortingAlgorithms/Code/MergeSort.py
# Merge Sort in Python
# Divide and conquer algorithm - divide array in half, sort, and merge

def merge(arr, left, mid, right):
    """Merge two sorted subarrays"""
    left_arr = arr[left:mid + 1]
    right_arr = arr[mid + 1:right + 1]
    
    i = j = 0
    k = left
    
    while i < len(left_arr) and j < len(right_arr):
        if left_arr[i] <= right_arr[j]:
            arr[k] = left_arr[i]
            i += 1
        else:
            arr[k] = right_arr[j]
            j += 1
        k += 1
    
    while i < len(left_arr):
        arr[k] = left_arr[i]
        i += 1
        k += 1
    
    while j < len(right_arr):
        arr[k] = right_arr[j]
        j += 1
        k += 1

def merge_sort(arr, left=0, right=None):
    """Divide and conquer merge sort"""
    if right is None:
        right = len(arr) - 1
    
    if left < right:
        mid = left + (right - left) // 2
        
        merge_sort(arr, left, mid)
        merge_sort(arr, mid + 1, right)
        merge(arr, left, mid, right)
    
    return arr

def merge_sort_copy(arr):
    """Non-in-place merge sort"""
    if len(arr) <= 1:
        return arr
    
    mid = len(arr) // 2
    left = merge_sort_copy(arr[:mid])
    right = merge_sort_copy(arr[mid:])
    
    return merge_helper(left, right)

def merge_helper(left, right):
    """Helper function to merge two sorted arrays"""
    result = []
    i = j = 0
    
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    
    result.extend(left[i:])
    result.extend(right[j:])
    return result

# Test 1: Basic merge sort
arr = [64, 34, 25, 12, 22, 11, 90]
print("Original Array:", arr)
result = merge_sort(arr.copy())
print("Sorted Array:", result)

# Test 2: Non-in-place version
arr2 = [64, 34, 25, 12, 22, 11, 90]
result2 = merge_sort_copy(arr2)
print("Sorted (Non-in-place):", result2)

# Test 3: Already sorted
arr3 = [1, 2, 3, 4, 5]
print("\nAlready sorted:", arr3)
result3 = merge_sort(arr3.copy())
print("After Merge Sort:", result3)

# Test 4: Reverse sorted
arr4 = [5, 4, 3, 2, 1]
print("\nReverse sorted:", arr4)
result4 = merge_sort(arr4.copy())
print("After Merge Sort:", result4)

# Test 5: Duplicates
arr5 = [5, 2, 8, 2, 9, 1, 5, 5]
print("\nArray with duplicates:", arr5)
result5 = merge_sort(arr5.copy())
print("After Merge Sort:", result5)

# Test 6: Single element
arr6 = [42]
print("\nSingle element:", arr6)
result6 = merge_sort(arr6.copy())
print("After Merge Sort:", result6)

print("\nComplexity Analysis:")
print("Time Complexity: O(n log n) - Linearithmic")
print("Space Complexity: O(n) - Requires temporary arrays")
print("Best Case: O(n log n)")
print("Worst Case: O(n log n)")
print("Average Case: O(n log n)")
print("Stable: Yes")
print("Note: Divide and conquer approach, consistent performance")

File: DSA/SortingAlgorithms/Code/MergeSort.cs
using System;
using System.Linq;

// Merge Sort in C#
// Divide and conquer algorithm - divide array in half, sort, and merge

class MergeSort {
    
    static void Merge(int[] arr, int left, int mid, int right) {
        int n1 = mid - left + 1;
        int n2 = right - mid;
        
        int[] leftArr = new int[n1];
        int[] rightArr = new int[n2];
        
        for (int i = 0; i < n1; i++)
            leftArr[i] = arr[left + i];
        for (int i = 0; i < n2; i++)
            rightArr[i] = arr[mid + 1 + i];
        
        int p = 0, q = 0, k = left;
        
        while (p < n1 && q < n2) {
            if (leftArr[p] <= rightArr[q]) {
                arr[k++] = leftArr[p++];
            } else {
                arr[k++] = rightArr[q++];
            }
        }
        
        while (p < n1)
            arr[k++] = leftArr[p++];
        
        while (q < n2)
            arr[k++] = rightArr[q++];
    }
    
    static void MergeSortMethod(int[] arr, int left, int right) {
        if (left < right) {
            int mid = left + (right - left) / 2;
            
            MergeSortMethod(arr, left, mid);
            MergeSortMethod(arr, mid + 1, right);
            Merge(arr, left, mid, right);
        }
    }
    
    static void DisplayArray(int[] arr) {
        Console.WriteLine(string.Join(" ", arr));
    }
    
    static void Main() {
        // Test 1: Basic merge sort
        int[] arr = {64, 34, 25, 12, 22, 11, 90};
        Console.WriteLine("Original Array: ");
        DisplayArray(arr);
        
        MergeSortMethod(arr, 0, arr.Length - 1);
        Console.WriteLine("Sorted Array (Merge Sort): ");
        DisplayArray(arr);
        
        // Test 2: Already sorted
        int[] arr2 = {1, 2, 3, 4, 5};
        Console.WriteLine("\nAlready sorted: ");
        DisplayArray(arr2);
        MergeSortMethod(arr2, 0, arr2.Length - 1);
        Console.WriteLine("After Merge Sort: ");
        DisplayArray(arr2);
        
        // Test 3: Reverse sorted
        int[] arr3 = {5, 4, 3, 2, 1};
        Console.WriteLine("\nReverse sorted: ");
        DisplayArray(arr3);
        MergeSortMethod(arr3, 0, arr3.Length - 1);
        Console.WriteLine("After Merge Sort: ");
        DisplayArray(arr3);
        
        // Test 4: Duplicates
        int[] arr4 = {5, 2, 8, 2, 9, 1, 5, 5};
        Console.WriteLine("\nArray with duplicates: ");
        DisplayArray(arr4);
        MergeSortMethod(arr4, 0, arr4.Length - 1);
        Console.WriteLine("After Merge Sort: ");
        DisplayArray(arr4);
        
        // Test 5: Single element
        int[] arr5 = {42};
        Console.WriteLine("\nSingle element: ");
        DisplayArray(arr5);
        MergeSortMethod(arr5, 0, arr5.Length - 1);
        Console.WriteLine("After Merge Sort: ");
        DisplayArray(arr5);
        
        Console.WriteLine("\nComplexity Analysis:");
        Console.WriteLine("Time Complexity: O(n log n) - Linearithmic");
        Console.WriteLine("Space Complexity: O(n) - Requires temporary arrays");
        Console.WriteLine("Best Case: O(n log n)");
        Console.WriteLine("Worst Case: O(n log n)");
        Console.WriteLine("Average Case: O(n log n)");
        Console.WriteLine("Stable: Yes");
        Console.WriteLine("Note: Divide and conquer approach, consistent performance");
    }
}
